### DISCUSSION TITLE
[Bug](undertow) Prepare for async repeatedly. Span is already in async mode. at org.apache.skywalking.apm.agent.core.context.trace.AbstractTracingSpan.prepareForAsync

### DISCUSSION THREAD

**Discussion by sutianbo** [apache/skywalking#9573] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

jdk version 11
skywalking version 8.7.0

java.lang.RuntimeException: Prepare for async repeatedly. Span is already in async mode. at org.apache.skywalking.apm.agent.core.context.trace.AbstractTracingSpan.prepareForAsync(AbstractTracingSpan.java:319) at org.apache.skywalking.apm.plugin.undertow.v2x.handler.TracingHandler.handleRequest(TracingHandler.java:68) at org.apache.skywalking.apm.plugin.undertow.v2x.handler.TracingHandler.handleRequest(TracingHandler.java:84) at io.undertow.server.Connectors.executeRootHandler(Connectors.java:387) at io.undertow.server.protocol.http.HttpReadListener.handleEventWithNoRunningRequest(HttpReadListener.java:256) at io.undertow.server.protocol.http.HttpReadListener.handleEvent(HttpReadListener.java:136) at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:162) at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:100) at io.undertow.server.protocol.http.HttpOpenListener.handleEvent(HttpOpenListener.java:57) at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92) at org.xnio.ChannelListeners$10.handleEvent(ChannelListeners.java:291) at org.xnio.ChannelListeners$10.handleEvent(ChannelListeners.java:286) at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92) at org.xnio.nio.QueuedNioTcpServer2.acceptTask(QueuedNioTcpServer2.java:178) at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:612) at org.xnio.nio.WorkerThread.run(WorkerThread.java:479)

### What you expected to happen

no exception

### How to reproduce

Any http Gateway request may be reproduce

### Anything else

org.apache.skywalking.apm.plugin.undertow.v2x.handler.TracingHandler
try {
            span.prepareForAsync();
            exchange.addExchangeCompleteListener(new ExchangeCompletionListener() {
                @Override
                public void exchangeEvent(HttpServerExchange httpServerExchange, NextListener nextListener) {
                    nextListener.proceed();
                    if (httpServerExchange.getStatusCode() >= 400) {
                        span.errorOccurred();
                        Tags.STATUS_CODE.set(span, Integer.toString(httpServerExchange.getStatusCode()));
                    }
                    span.asyncFinish();
                }
            });
        } catch (Throwable e) {
            ContextManager.activeSpan().log(e);
        }
        try {
            next.handleRequest(exchange);
        } catch (Throwable e) {
            span.log(e);
        } finally {
            ContextManager.stopSpan(span);
            ContextManager.getRuntimeContext().remove(Constants.FORWARD_REQUEST_FLAG);
        }
        
        It could be span.asyncfinish (); Is in asynchronous, not yet executed, but executes next. HandleRequest (exchange);

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Agent plugins are contributed by community. We would not debug for you. You have to debug by yourself.

**Comment by wu-sheng** [comment]
Also, 8.7.0 is a very old release. You could try 8.12.0

**Reply by zhang-zhipeng** [reply]
I have the same problem.
environment
     jdk version 8、skywalking version 9.2.0、java-agent 8.12.0、undertow、spring-cloud-gateway
when the span in TracingHandler calls prepareForAsync.isInAsyncMode = true, so when the span in DispatcherHandlerHandleMethodInterceptor calls prepareForAsync again, the check will be triggered throw new RuntimeException("Prepare for async repeatedly. Span is already in async mode.");

![call stack](https://user-images.githubusercontent.com/42615740/190459147-102039c2-d759-4496-aa3e-289912aa0e61.jpg)
![DispatcherHandlerHandleMethodInterceptor](https://user-images.githubusercontent.com/42615740/190459156-2022f5a7-5d1e-42d1-8209-965b03f34fd7.jpg)
![TracingHnadler](https://user-images.githubusercontent.com/42615740/190459157-a9002481-d1e8-4a56-a49d-a11d3daeddf7.jpg)

**Reply by chengchaoyin** [reply]
I have the same problem.
Was it finally solved？

**Reply by zhang-zhipeng** [reply]
remove the undertow plugin，use netty-web-server.

**Comment by lbrant** [comment]
What is the problem now, and how can it be solved？
